### DISCUSSION TITLE
Add intellisense and more autocompletion

### DISCUSSION THREAD

**Discussion by ross96D** [a-h/templ#305] [Ideas]
I think it would be productive for templ users if the html had intellisense, so that they could have information about the tags and attributes. I think this is quite easy to do, I was able to do a minimum example just by making small changes to the library and adding the definitions as a json that I got from https://github.com/microsoft/vscode-html-languageservice/blob/main/src/languageFacts/data/webCustomData.ts . As in the photo below

![image](https://github.com/a-h/templ/assets/79549577/fe0cdffb-4dff-473a-a4f3-e691c377d7fe)

About autocompletion I know there are snippets but I think this approach has some cons. For example, it would be good if when I put an attribute with autocompletion it would automatically add the attribute plus the equals and the double quotes. And that the autocomplete brings the definitions of the attribute/tag that corresponds to it.

In this case I don't know how to achieve it and I don't know how difficult it would be to do it.

**Comment by a-h** [comment]
Totally agree! There was a bit of chat about it in this issue: https://github.com/a-h/templ/pull/244#issuecomment-1774752975

In there, I've dropped a few links in around how the integration might work.

At this point, I don't know whether it's something that should be done within the `templ lsp` command, or within the `templ-vscode` and `templ-goland` extensions - i.e. I think that since editors support having multiple LSPs per file, maybe it's possible to use the editor's built-in functionality to handle the HTML parts of templ.

But... that might not work, so `templ lsp` might need to:

* Probe for the existence of a HTML LSP in the path and start it up if it exists.
* Identify when the user is in a HTML part of templ, and proxy through the command to the HTML LSP.
  * So that the HTML LSP doesn't get confused by the templ code (i.e. the Go code), these parts of the file would be replaced with spaces.

I just haven't had chance to do much further research into it yet...

**Comment by a-h** [comment]
Oh, in my Gophercon UK talk from earlier this year, I talk about how the underlying protocol works, and cover how it works with templ. https://youtu.be/EkK8Jxjj95s

**Comment by ross96D** [comment]
Hey, sorry for the delay, I haven't had time to deeply look into this solutions. Just saw the vscode request forwarding and it seems like a nice solution, but since i have a lack of knowledge on nvim and goland I don't really know how feasible it is.

The thing with the templ lsp having the html lsp is that it requires more maintenance and will be bit bloated.

Assuming the request forwarding as solution, maybe is worth at least that the templ cli can identify when the user is in a HTML part of templ, so the client don't have to rewrite this code.

PD: I saw your talk like 1-2 months ago.. great talk :100: .
